### PR TITLE
[GE] Parenting System Changes

### PR DESCRIPTION
**Pull request by msDestiny14**
- Reorder components with drag and drop.
- Visual changes (shifting)
- Bug fixes

### CODE REVIEW

File: guiEditor/src/diagram/workbench.tsx [hunk 1]
```diff
@@ -294,19 +294,42 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
...
+            draggedControlParent.removeControl(draggedControl);
+            
+            if (control != null) {
+                if (this.props.globalState.workbench.isContainer(control)) {
+                    this.props.globalState.guiTexture.removeControl(draggedControl);
+                    (control as Container).addControl(draggedControl);
+                }
+                if (control.parent) {
+                    let index = control.parent.children.indexOf(control);
+                    index = this._adjustParentingIndex(index);
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:308]: This looks like black magic :-)

File: guiEditor/src/tools.ts [hunk 2]
```diff
@@ -41,16 +41,17 @@ export class Tools {
         if (parent && parent.reservedDataStore && parent.reservedDataStore.detachedChildren) {
             finalArray.push(...parent.reservedDataStore.detachedChildren);
         }
+        return finalArray.reverse();
 
         return finalArray.sort((a: any, b: any) => {
```
Comment (sebavan) [inline guiEditor/src/tools.ts:46]: This is useless if you return 2 lines before

**Review by sebavan — CHANGES_REQUESTED** [review]

File: guiEditor/src/tools.ts (hunk 2 above)
Comment (msDestiny14) [reply guiEditor/src/tools.ts:46]: Aha forgot to remove ;) Good find.

File: guiEditor/src/components/sceneExplorer/treeItemComponent.tsx [hunk 3]
```diff
@@ -27,8 +30,7 @@ class TreeItemExpandableHeaderComponent extends React.Component<ITreeItemExpanda
     }
 
     render() {
-        const chevron = this.props.isExpanded ? <FontAwesomeIcon icon={faMinus} /> : <FontAwesomeIcon icon={faPlus} />
-        const expandAll = this.props.isExpanded ? <FontAwesomeIcon icon={faCompress} /> : <FontAwesomeIcon icon={faExpandArrowsAlt} />
+        const chevron = this.props.isExpanded ? <img src={expandedIcon} className="icon"/>: <img src={collapsedIcon} className="icon"/>
```
[the remark this answers is not in the document]
Comment (msDestiny14) [reply guiEditor/src/components/sceneExplorer/treeItemComponent.tsx:33]: It's used more than once.

File: guiEditor/src/diagram/workbench.tsx [hunk 4]
```diff
@@ -294,19 +294,42 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
     private parent(control: Nullable<Control>) {
         const draggedControl = this.props.globalState.draggedControl;
 
-        if (draggedControl != null) {
-            if (draggedControl.parent) {
-                (draggedControl.parent as Container).removeControl(draggedControl);
+        if (draggedControl != null && draggedControl.parent) {
+            let draggedControlParent = draggedControl.parent;
```
[the remark this answers is not in the document]
Comment (msDestiny14) [reply guiEditor/src/diagram/workbench.tsx:298]: While I love this, compiler is not letting me do this

**Review by sebavan — APPROVED** [review]

**Pull request merged**